Add release notes for GOV.UK Prototype Kit v 9.15.0 (#1106)
alphagov/govuk-prototype-kit: CHANGELOG.md

# Unreleased

# 9.15.0 (Feature release)

## New features

### Replace back link placeholder URLs with JavaScript

We've added JavaScript to make the [back link component](https://design-system.service.gov.uk/components/back-link/) take users to the previous page by default, rather than you having to update placeholder text.

> If you cannot rely on JavaScript always being available for production scenarios, do not use this JavaScript-reliant approach.

Reviewer: Sorry @PippaClarkGDS, my bad - I think we might need to tweak this sentence. As @vanitabarrett pointed out, we know people should not rely on JavaScript being available in production, as some users will turn it off or disable it. So, it's not really an if'.

This feature helps people who are quickly building prototypes (which might not be 100% accessible). We want to advise against them simply re-using this code that isn't suitable for production scenarios.

Could the following work?

```suggestion
Do not use this JavaScript-reliant approach in production, as it excludes users who may not have JavaScript available
```

Reviewer: @vanitabarrett Just chatted with @PippaClarkGDS about this. How's this wording?

```suggestion
If you want to use JavaScript in production, you must also use a non-JavaScript alternative, so you do not exclude any users.
```

Reviewer: @PippaClarkGDS As agreed, I'll add in this wording:

```suggestion
If you want to use JavaScript in production, you must also use a non-JavaScript alternative, so you do not exclude any users. If you cannot use a non-JavaScript alternative, you should hide the back link when JavaScript is not available.
```

# Unreleased

> ## New features

Reviewer: ```suggestion
# # 9.15.0 (Feature release)

## New features
```

### Replace back link placeholder URLs with JavaScript

> We've added JavaScript to make the [back link component]( https://design-system.service.gov.uk/components/back-link/) take users to the previous page by default.

Reviewer 2: ```suggestion
We've added JavaScript to make the [back link component](https://design-system.service.gov.uk/components/back-link/) take users to the previous page by default.
```

Reviewer: D'oh my bad, thanks for flagging

> You can now use JavaScript to make the [back link component]( https://design-system.service.gov.uk/components/back-link/) take users to the previous page by default.

Reviewer 2: @EoinShaughnessy I don't think this quite captures the change - we've made the back link in the kit use JavaScript *by default* so that it will 'just work' without users needing to do anything. Before, we had some placeholder text which users had to manually update to get the back link in their prototypes to work. This means they no longer need to do that.

Reviewer: Ah I see - thanks. What about something like this?

```suggestion
We've added JavaScript to make the [back link component]( https://design-system.service.gov.uk/components/back-link/) take users to the previous page by default.
```

Reviewer 2: @EoinShaughnessy That's great! 🎉

> Before, you had to manually replace the `href` attribute for each back link in the page templates.

Reviewer: ```suggestion
Before, to get your prototypes to work, you had to manually update the placeholder text in the back link.
```

Reviewer 2: Yep!

Reviewer: Cheers!

Do not use this approach for production scenarios, where you cannot rely on JavaScript always being available.

> You can still override the `href` attribute if you need different behaviour.

Reviewer: @lfdebrux @vanitabarrett Should we give a 'for example'? Or would that be overkill/unnecessary for some reason?

Reviewer 2: Hm, not sure about this. Our guidance in the design system says "Make sure the link takes users to the previous page they were on, in the state they last saw it." so I don't think we should encourage people to change the behaviour of what the back link does. They may need to override the `href` to provide a solution which works when JavaScript is disabled though/

Reviewer: Ah interesting. Should we change to something like this?

```suggestion
You can still override the `href` attribute if you need to provide a solution that works when JavaScript is disabled.
```

Reviewer 2: Sounds good 👍

This was added in [Pull request #1103: Replace back link placeholder URLs with JavaScript](https://github.com/alphagov/govuk-prototype-kit/pull/1103)

### Update Node.js to protect your code

If you use the GOV.UK Prototype Kit, then we recommend you update to the latest version of Node.js 14 Long Term Support (LTS).

This is to make sure you're protected from a [recent security vulnerability in the npm (Node Package Manager)](https://github.blog/2021-09-08-github-security-update-vulnerabilities-tar-npmcli-arborist/).

> To make sure you are using Node.js version 14.17.6 or later, follow the [install instructions on the Prototype Kit website](https://govuk-prototype-kit.herokuapp.com/docs/install/requirements.md#nodejs-version-14-lts).

Reviewer: ```suggestion
To make sure you're using Node.js version 14.17.6 or later, follow the [install instructions on the Prototype Kit website](https://govuk-prototype-kit.herokuapp.com/docs/install/requirements.md#nodejs-version-14-lts).
```

> If you are using Node Version Manager (nvm), you can instead [run `nvm install` to install v14.17.6](https://github.com/alphagov/govuk-prototype-kit/pull/1076).

Reviewer: ```suggestion
If you're using Node Version Manager (nvm), you can instead [run `nvm install` to install v14.17.6](https://github.com/alphagov/govuk-prototype-kit/pull/1076).
```

# Unreleased

> # # 9.15.0 (Feature release)

Reviewer: ```suggestion
# 9.15.0 (Feature release)
```

## New features

### Replace back link placeholder URLs with JavaScript

> We've added JavaScript to make the [back link component](https://design-system.service.gov.uk/components/back-link/) take users to the previous page by default.
> 
> Before, to get the back link to work, you had to manually update placeholder text.

Reviewer 3: ```suggestion
We've added JavaScript to make the [back link component](https://design-system.service.gov.uk/components/back-link/) take users to the previous page by default, rather than you having to update placeholder text.
```

Reviewer 3: blending these 2 as the 'Before' line sounded a little clunky out of context

> Do not use this approach for production scenarios, where you cannot rely on JavaScript always being available.

Reviewer 4: Can we change to "Do not use the back link component in production"? The previous two sentences refer to two different things - using the back link component with javascript and using the back link component with manually updated placeholder text. So I think it would help to be explicitly clear.

Reviewer: That's a great point, thanks Jon.

```suggestion
Do not use the back link component in production, where you cannot rely on JavaScript always being available.
```

Reviewer 2: I think this implies the back link component should never be used in production - what we're really saying is that this approach (which relies on JavaScript with no fallback) shouldn't be used.

Reviewer: @jonathanglassman @vanitabarrett What about something like this?

```suggestion
Do not use JavaScript for production scenarios, where you cannot rely on JavaScript always being available.
```

Reviewer 2: This is tricky to word, I'm still not sure if it's a bit misleading because users *can* use JavaScript. In our guidance in the design system we say:

> Where possible, ensure it works even when JavaScript is not available.
> If this is not possible, you should hide the back link when JavaScript is not available.

Reviewer 2: Maybe something like "Do not rely on JavaScript for production scenarios..." or "Do not rely on JavaScript-only implementations for production scenarios..." is clearer?

Reviewer: "Implementations" is specific about what to use JavaScript for, which sounds like it could help users. What do you think, @jonathanglassman? 

```suggestion
Do not rely on JavaScript-only implementations for production scenarios, where you cannot rely on JavaScript always being available.
```

Reviewer 4: How about "Do not use this new Javascript-reliant approach for production scenarios, where you cannot rely on JavaScript always being available."

Reviewer: I think that sounds good. Is this wording ok, @vanitabarrett?

```suggestion
Do not use this JavaScript-reliant approach for production scenarios, where you cannot rely on JavaScript always being available.
```

Reviewer 2: Sounds good, thanks both!

Reviewer: Thanks @Vanita! Thanks @jonathanglassman!

> Do not use this JavaScript-reliant approach for production scenarios, where you cannot rely on JavaScript always being available.

Reviewer 3: ```suggestion
If you cannot rely on JavaScript always being available for production scenarios, do not use this JavaScript-reliant approach.
```

Reviewer 3: small tweak to front-load this sentence

You can still override the `href` attribute if you need to provide a solution that works when JavaScript is disabled.

This was added in [Pull request #1103: Replace back link placeholder URLs with JavaScript](https://github.com/alphagov/govuk-prototype-kit/pull/1103)

### Update Node.js to protect your code

> If you use the GOV.UK Prototype Kit, then we recommend you update to the latest version of Node.js 14 Long Term Support (LTS).

Reviewer 3: ```suggestion
If you use the GOV.UK Prototype Kit, we recommend you update to use the latest version of Node.js 14 Long Term Support (LTS).
```

> This is to make sure you're protected from a [recent security vulnerability in the npm (Node Package Manager)](https://github.blog/2021-09-08-github-security-update-vulnerabilities-tar-npmcli-arborist/).

Reviewer 3: I would have this follow on from last line rather than separate line

> To make sure you're using Node.js version 14.17.6 or later, follow the [install instructions on the Prototype Kit website](https://govuk-prototype-kit.herokuapp.com/docs/install/requirements.md#nodejs-version-14-lts).

Reviewer 3: ```suggestion
To make sure you're using Node.js version 14.17.6 or later, [follow the install instructions on the Prototype Kit website](https://govuk-prototype-kit.herokuapp.com/docs/install/requirements.md#nodejs-version-14-lts).
```

## Fixes

- [Pull request #1104: Change visually hidden footer title](https://github.com/alphagov/govuk-prototype-kit/pull/1104)

> ### Updating Node.js

Reviewer: Maybe we could say why users should do this. So, something like:

```suggestion
### Update Node.js to protect your code
```

> We are recommending that all users of the GOV.UK Prototype Kit update to the latest version of Node.js 14 LTS, to make sure they are protected from a [recent security vulnerability in the Node package manager npm](https://github.blog/2021-09-08-github-security-update-vulnerabilities-tar-npmcli-arborist/).

Reviewer: ```suggestion
If you use the GOV.UK Prototype Kit, then we recommend you update to the latest version of Node.js 14 Long Term Support (LTS).

This is to make sure you're protected from a recent security vulnerability in the Node Package Manager (npm).
```

Author: Thanks, this reads much better! It does remove the link though, is that on purpose?

Reviewer: No, that's my mistake! Ta for flagging. Will add back in.

```suggestion
If you use the GOV.UK Prototype Kit, then we recommend you update to the latest version of Node.js 14 Long Term Support (LTS).

This is to make sure you're protected from a [recent security vulnerability in the npm (Node Package Manager)](https://github.blog/2021-09-08-github-security-update-vulnerabilities-tar-npmcli-arborist/).
```

> Please follow the [instructions on the prototype kit website](https://govuk-prototype-kit.herokuapp.com/docs/install/requirements.md#nodejs-version-14-lts) to make sure you are using Node.js version 14.17.6 or later.

Reviewer: ```suggestion
To make sure you are using Node.js version 14.17.6 or later, follow the [install instructions on the prototype kit website](https://govuk-prototype-kit.herokuapp.com/docs/install/requirements.md#nodejs-version-14-lts).
```

> If you are using Node Version Manager (nvm), you can if you wish instead [run `nvm install` to install v14.17.6](https://github.com/alphagov/govuk-prototype-kit/pull/1076).

Reviewer: ```suggestion
If you are using the Node Version Manager (nvm), you can instead [run `nvm install` to install v14.17.6](https://github.com/alphagov/govuk-prototype-kit/pull/1076).
```

Author: Small nitpick, but I think with nvm we should either go with 'the Node version manager `nvm`,' or 'Node Version Manager' without the 'the'. Writing 'the Node Version Manager' looks wrong to me, because it's the proper name of the script.

Reviewer: Ah gotcha, yeah that makes sense.

```suggestion
If you are uNode Version Manager (nvm), you can instead [run `nvm install` to install v14.17.6](https://github.com/alphagov/govuk-prototype-kit/pull/1076).
```
